### PR TITLE
fix: resolve card-facing packages in published boxel-cli parse

### PR DESCRIPTION
**Pull request by habdelra**
## What

`boxel parse` type-checks card code in a temp workspace whose `node_modules` is assembled by symlinking boxel-cli's declared dependencies. The set of modules card code may actually import is a different, larger set: the one the host shims onto the virtual network in `packages/host/app/lib/externals.ts`.

Where those two lists disagree, a published install reports `Cannot find module` — or, for a package that resolves without declarations, `implicitly has an 'any' type` — against card code that is correct and runs fine. The monorepo hides this: that branch symlinks host's entire `node_modules`, so the failure only appears once the CLI is installed from npm.

In a fresh `npm i @cardstack/boxel-cli` project, `ember-modifier`, `ember-concurrency`, `@ember/test-helpers`, `tracked-built-ins`, `date-fns` and `lodash-es` all failed to resolve. `packages/base` and `packages/experiments-realm` already ship cards importing them — 20 files each for `ember-modifier` and `ember-concurrency`, 13 for `lodash-es`, 8 for `tracked-built-ins`, 6 for `date-fns`. The worst outcome isn't the noise; it's that a phantom error invites an author to rewrite code that was never broken.

## How

- **Declare the shimmed npm packages** as boxel-cli dependencies at the versions host pins, plus the `@types/*` for those that ship no declarations of their own (`lodash`, `lodash-es`, `rsvp`, `flat`, `uuid`).
- **Pin `ember-source`.** `@ember/*` types are a large share of the card-facing surface and were reaching published installs only as an unpinned transitive peer, resolving to 7.x while host runs ~6.10 — so parse checked card code against a different major of Ember than the runtime.
- **Bundle `@cardstack/bxl`** into `bundled-types/` behind a path alias, the way `runtime-common` already is. It's a workspace package, so a runtime dependency would couple boxel-cli's publish to bxl's. Its `exports` resolve to `.ts`, so the vendored source *is* the type surface — a wrong formula argument is a real diagnostic rather than a module silently resolved as `any`. `allowImportingTsExtensions` keeps its explicit `.ts` specifiers from emitting TS5097s that would leave `totalDiagnosticLines` permanently non-zero and disarm the "glint resolved nothing" check.
- **Point both `@cardstack/boxel-host` alias spellings at `tools/`.** The host registers each tool module under its `tools/` specifier and the pre-rename `commands/` one; the latter aliased a directory that no longer exists.
- **Cover `ember-source/types`** with an ambient declaration matching the empty stub the host shims — its `./types` entry resolves to an ambient global rather than a module, and `./types/preview` has no export entry at all.

## Keeping the two lists together

Declaring the packages fixes today's gap; on its own it would drift again. `tests/card-runtime-packages.test.ts` resolves every shimmed specifier **through TypeScript itself** and requires the result to be a declaration file. Membership in `package.json` is the weaker property — it is satisfied by a package that ships no types at all, which is exactly how `flat` and `uuid` stayed broken after being declared.

Specifiers that legitimately don't resolve through `node_modules` are accounted for explicitly: a path alias, glint's Ember environment, a vendored ambient declaration in `host/types`, or a generated shim. The `@ember/*` allowance is an allowlist rather than a prefix rule, because `@ember/string` and `@ember/test-waiters` are ordinary packages sharing that scope. A companion assertion fails when an allowance outlives the specifier it was written for, and another fails when a shim call site is written in a form the specifier patterns can't read — an identifier, a template literal, a `{ prefix }` descriptor — so the suite can't silently check less than it claims.

## Test plan

- `pnpm lint:js`, `pnpm lint:types` and `pnpm test:unit-exclude-smoke` (580 tests) pass in `packages/boxel-cli`.
- The guard fails as intended on each drift mode, verified by mutation: an undeclared shimmed package, a declared-but-untyped one, and a shim call site written as a `{ prefix }` descriptor.
- Two new fixtures join `CLEAN_FIXTURES`, so they run in the tarball/published contexts the parse suite already covers: `card-runtime-imports` (a card and a `.test.gts` using the previously-unresolvable imports) and `bxl-formula` (a `computeVia` formula).
- Verified end to end by packing this branch and `npm install`ing the tarball into a plain npm project (fully hoisted, no nested `node_modules`): **all 77 shimmed specifiers resolve with real declarations**, both new fixtures and all six existing clean fixtures pass, and the two negative fixtures still produce their real diagnostics. `ember-source` resolves to 6.10.1 there, matching host.

## Notes for review

- The install grows 237 MB → 282 MB. `date-fns` (34 MB), `ember-source` (18 MB), `ethers` (17 MB) and `matrix-js-sdk` (14 MB) are most of it. Real packages are the only mechanism that gives real types here, since `linkResolvedDeps` reads `dependencies`. If that trade is wrong for the heavy runtime-only shims, `ethers` and `matrix-js-sdk` could take the ambient-declaration treatment instead — still strictly better than `Cannot find module`.
- **One gap this cannot close:** `patchedDependencies` applies in the monorepo only. `matrix-js-sdk`'s patch widens its event interfaces, so a card typing an `app.boxel.*` matrix event checks clean in the monorepo and errors against a published install's unpatched declarations. Noted in `parse.ts` rather than worked around.
- The lockfile diff is large, but every resolved-version change is confined to `packages/boxel-cli`. No other importer's versions move; the rest is pnpm recomputing peer-context keys.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

https://claude.ai/code/session_01V5chbxveQrAru4pXuJb7bz

### CODE REVIEW

File: packages/boxel-cli/scripts/build-types.ts [hunk 1]
```diff
@@ -545,6 +554,16 @@ const VENDORS: Vendor[] = [
...
   },
+  {
+    // `@cardstack/bxl` is card-facing (a card's `computeVia` formula
+    // imports it) and its package `exports` resolve to `.ts` source, so
+    // the type surface is the source itself — no `.d.ts` emit needed.
+    // Reached through the `@cardstack/bxl` path alias in parse.ts.
+    name: 'bxl',
+    from: join(MONOREPO_PACKAGES, 'bxl', 'src'),
+    to: join(PACKAGE_ROOT, 'bundled-types', 'bxl'),
+    filter: skipMonorepoArtifacts,
```
[the remark this answers is not in the document]
Comment (habdelra) [reply packages/boxel-cli/scripts/build-types.ts:565]: [Claude Code 🤖]

Confirmed and fixed in `d01590f4` — `packages/bxl/src/**` is now a publish trigger for the CLI.

Verified the premise: `boxel-cli-publish.yml` watched only `packages/boxel-cli/**`, so a BXL-only change left `bundled-types/bxl` frozen in the published CLI while the host advanced.

One correction to the other half of the finding: CI change-detection already covers this. The `realm-server` filter includes `packages/bxl/**`, and `boxel-cli-test` runs on `realm-server == 'true'`, so the parse fixtures already ran on BXL changes. (The `boxel-cli` filter lines the comment cites no longer exist — that entry was removed earlier in this branch as redundant for the same reason.) So only the release trigger needed widening.

I did not take the "consume the published BXL type surface instead" option: a `workspace:*` dependency would couple the CLI's publish to BXL's, so an `npm i @cardstack/boxel-cli` could resolve a `@cardstack/bxl` version that isn't on npm yet.

Worth flagging that this staleness is a general property of `bundled-types`, not specific to BXL: the CLI also vendors from `packages/base`, `runtime-common`, `host/{app,tests,types}`, `boxel-ui`, `local-types` and `skills-realm`, none of which trigger a release either. That's pre-existing, and widening the trigger to `packages/host/**` or `packages/base/**` would cut an unstable release on nearly every PR — a release-cadence decision I've left to the team rather than making unilaterally. Happy to follow up if you want the rest covered.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: packages/boxel-cli/scripts/build-types.ts [hunk 2]
```diff
@@ -639,6 +658,21 @@ declare module '@cardstack/boxel-icons/*' {
...
+
+// \`ember-source\` exposes no \`./types\` entry through its package
+// \`exports\`, so these specifiers resolve for card code only because the
+// host shims them onto the virtual network as empty stubs. Mirror that
+// stub shape here so an import type-checks to the same nothing it
+// evaluates to at runtime.
+declare module 'ember-source/types' {
+  const value: any;
+  export default value;
+}
```
[the remark this answers is not in the document]
Comment (habdelra) [reply packages/boxel-cli/scripts/build-types.ts:671]: [Claude Code 🤖]

The premise is right — `packages/local-types/index.d.ts:1` really is a bare `import 'ember-source/types'`, and an ambient `declare module` for that specifier does satisfy it — but the predicted consequence doesn't reproduce. Ember's real declarations still load.

Tested against a packed-and-npm-installed tarball of this branch (fully hoisted), with a probe that distinguishes real declarations from a permissive stub — a bogus named import errors only if the module's true types are loaded:

```ts
import { get, thisExportDoesNotExist } from '@ember/object';
import { on } from '@ember/modifier';
```

| shims | result |
| --- | --- |
| as shipped (ambient `ember-source/types` present) | `TS2305: Module '"@ember/object"' has no exported member 'thisExportDoesNotExist'` |
| ambient `ember-source/types` declarations stripped | identical `TS2305` |

Byte-identical either way, so the stub isn't displacing anything — `@ember/*` types arrive through glint's environment (`@glint/ember-tsc/types`, imported two lines below in the same file) rather than through that side-effect import. A sweep of all 77 specifiers `shimExternals` exposes, `@ember/owner` included, resolves clean on this head.

Two things did change since the commit reviewed here, both relevant: `ember-source` is now a pinned dependency, so it's linked into the parse workspace rather than absent, and the stub's stated rationale was wrong in the way you'd expect from the code — ember-source *does* expose a `./types` entry; it resolves to an ambient global declaration file rather than a module, which is the actual reason the specifier needs help. That comment has been corrected.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: packages/boxel-cli/tests/card-runtime-packages.test.ts [hunk 3]
```diff
@@ -0,0 +1,249 @@
...
+    }
+  }
+  return specifiers;
+}
+
+/** Every shim call site, however its specifier is written. */
+function countShimCallSites(): number {
+  let source = readFileSync(EXTERNALS_PATH, 'utf8');
+  return (source.match(/\.shim(?:Async)?Module\(/g) ?? []).length;
+}
```
[the remark this answers is not in the document]
Comment (habdelra) [reply packages/boxel-cli/tests/card-runtime-packages.test.ts]: [Claude Code 🤖]

Correct, and fixed in `5932b5b5` — the scan now walks the AST.

Verified the exact case you named: `externals.ts:73` explains the live-test override with a prose `loader.shimModule('qunit', QUnit)`, and the text scan was matching it. The AST finds **77** real call sites where the text scan found 78, so the extra one was that comment.

Worth noting the guard was passing for the wrong reason rather than merely being weakened: both regexes matched the comment, so the counts balanced and the `qunit` duplicate looked like a legitimate double shim. A package named only in prose would have been demanded as a dependency.

The AST walk also reads `shimAsyncModule`'s descriptor properly, which made the "every call site names a module" assertion precise: it now fails on the shapes that can't be followed to a package name, rather than on a count that happens to balance. Mutation-tested all four cases against the new version:

| mutation | result |
| --- | --- |
| a comment mentioning `loader.shimModule('totally-fake-pkg', x)` | ignored ✓ |
| `shimAsyncModule({ prefix: 'ns/', … })` | fails, naming the call site |
| a template-literal id | fails, naming the call site |
| a real shim of an undeclared package | fails, naming the package |

Lint, typecheck and the 580-test unit suite pass on the new head.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by habdelra** [comment]
[Claude Code 🤖]

**Host Tests (9, 16)** failed on `5932b5b5`. I don't believe it belongs to this PR, and here is what that rests on rather than an assertion that it's a flake.

**What this branch changes.** Ten files: `packages/boxel-cli/**`, `.github/workflows/boxel-cli-publish.yml`, and `pnpm-lock.yaml`. No file under `packages/host/` is touched, and `packages/host` declares no dependency on `@cardstack/boxel-cli`, so nothing in the host's test build reads any of it.

**The lockfile, which is the only plausible cross-package link.** I diffed every importer's resolved versions between `origin/main` and this head with peer-context suffixes stripped:

```
boxel-cli changes: 30   changes in every other package: 0
```

Every moved version is under `packages/boxel-cli`. `packages/host` resolves byte-identical dependencies to `main`, so the host bundle under test is built from the same graph. The rest of the lockfile diff is pnpm recomputing peer-context keys, which changes no installed version.

**What the job log shows.** No assertion failure reaches the console — the QUnit output goes to the `testem-log-9` artifact, which I can't fetch from here (the Azure blob host is blocked by this environment's egress proxy). What the console does carry, during the test window, is connect timeouts against a service this diff doesn't touch:

```
Failed to probe remote realm visibility for http://test-realm/test/pet [TypeError: fetch failed]
  [cause]: ConnectTimeoutError: Connect Timeout Error (attempted address: test-realm:80, timeout: 10000ms)
    code: 'UND_ERR_CONNECT_TIMEOUT'
```

**What I could not establish.** I could not compare against `main`: the recent `main` runs either were cancelled or didn't exercise the host suite (the last one was a grafanactl change). So "red on the base branch too" is untested, not ruled in.

**Next step.** I tried to re-run the failed job and GitHub refused with `This workflow is already running` — the other 15 host shards and the realm-server shards are still going. I'll re-run this shard once the run finishes, which is the one re-run I'll spend on it. If it fails again on the same commit I'll treat it as real and root-cause it rather than re-running further.

Everything else is green, including **Boxel CLI Tests** and **Boxel CLI Build** — the checks that actually exercise this change, the latter running the new guard suite and the parse fixtures against a packed-and-installed tarball.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by habdelra** [comment]
[Claude Code 🤖]

Re-ran **Host Tests (9, 16)** on the same commit, as said above — it passed (291 tests, 291 ✅). That was the one re-run; I haven't spent another.

Two things from the aggregate results support reading the first attempt as environmental rather than a test regression. The host suite reported `tests_fail: 0, tests_error: 1` — across 4603 tests not one assertion failed; a single *run* errored — and Realm Server was 2517/2517 green on the same commit throughout.

The head is now green across all 74 checks and `mergeable_state` is `clean`.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Review by lukemelia — APPROVED** [review]
